### PR TITLE
UI integration testing

### PR DESCRIPTION
**Pull request by noahdude1203**

### CODE REVIEW

File: package.json
Comment (radiantBear) [inline package.json:1]: Just curious, what are `sinon` and `xmlchars` being used for?

File: src/line/window.js
Comment (radiantBear) [inline src/line/window.js:1]: What's this being used for? I don't think we should be exporting a `window` global; can we undo this change?

File: src/line/i-chartBuilder.test.js [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+    //make sure the correct button is selected
+    expect(addButton).toHaveTextContent("+")
+
+    //count all cells
+    var xValues = await domTesting.getAllByLabelText(document, "X")
+    var yValues = await domTesting.getAllByLabelText(document, "Y")
+
+
+    // Assert that the lists are not empty
+	expect(xValues).toHaveLength(1)
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js]: There's some strange indentation on line 48; could you make sure it's indented with 4 spaces like the rest of the code?

File: src/line/i-chartBuilder.test.js [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+ * @jest-environment jsdom
+ */
+
+const fs = require("fs")
+const domTesting = require('@testing-library/dom')
+require('@testing-library/jest-dom')
+const userEvent = require("@testing-library/user-event").default
+
+//Spy on
+const window = require("./window")
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js]: Wait, isn't the `window` object supposed to be created by `jsdom`? I don't think we need to `require()` it directly.

File: src/line/i-chartBuilder.test.js [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+    jest.isolateModules(function() {
+        require(jsPath)
+    })
+}
+
+/*
+    They should verify the behaviors described above. Note that these behaviors are 
+    duplicated across the three chart builder pages (bar.html, line.html, and scatter.html). 
+    You only need to test these behaviors on one of those pages.
+*/
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js]: Could we remove this comment now that the tests have been written?

File: src/line/i-chartBuilder.test.js [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+    const user = userEvent.setup()
+
+    await user.click(addButton)
+
+    xValues = domTesting.getAllByLabelText(document, "X")
+    yValues = domTesting.getAllByLabelText(document, "Y")
+
+    expect(xValues).toHaveLength(2)
+    expect(yValues).toHaveLength(2)
+})
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:49]: The instructions for "Adding values in the chart builder" say 
> Clicking the button should not impact any data the user has already entered.

It looks like we need a second (or expanded) test to verify that behavior.

File: src/line/i-chartBuilder.test.js [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+test('Pressing "+" adds exactly one empty x and y value box', async function(){
+    initDomFromFiles(`${__dirname}/line.html`, `${__dirname}/line.js`)
+
+    //get + button for adding  new x and y cells.
+    const buttons = domTesting.queryAllByRole(document, "button")
+    const addButton = buttons[3];
+    
+    //make sure the correct button is selected
+    expect(addButton).toHaveTextContent("+")
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js]: I'm not in love with this method for getting buttons. I see how this verifies that the test is clicking the button it expects to, but it's still a brittle solution. If the ordering of the buttons changes for any reason, this assertion would cause the test to fail even if the SUT functions properly.

I think this should be changed throughout the integration tests to instead use a `.getByText()` query, maybe using the optional parameters for that function to ensure that the selected element is a button if you'd like to ensure that.

File: src/line/i-chartBuilder.test.js [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+    await user.type(xLabel, "X Label Name :)")
+    await user.type(yLabel, "Y Label Name :(")
+    await user.type(xValue, "22")
+    await user.type(yValue, "33")
+
+    await user.click(generateButton)
+    
+    expect(alertMock).toHaveBeenCalledTimes(1)
+})
+
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:118]: It looks like there's no assertion that an empty X label results in an alert. Is there any particular reason that wasn't added?

File: src/line/i-chartBuilder.test.js [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+    await user.type(xLabel, "X Label Name :)")
+    await user.type(yLabel, "Y Label Name :(")
+    await user.type(xValue, "22")
+    await user.type(yValue, "33")
+
+    await user.click(generateButton)
+    
+    expect(alertMock).toHaveBeenCalledTimes(1)
+})
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:117]: To preserve test isolation, `alertMock.mockRestore()` and `window.alert.mockRestore()` should be called at the end of or between each test. Am I missing where that happens, or does that need to be added?

File: src/line/i-chartBuilder.test.js [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+    const yValue = yValues[0]
+
+    const buttons = domTesting.queryAllByRole(document, "button")
+    var generateButton = buttons[0]
+
+    //make sure the correct button is selected
+    expect(generateButton).toHaveTextContent("Generate chart")
+
+    //Popups
+    window.alert = jest.fn();
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:101]: Not sure it matters a whole ton, but does `window.alert` need to be mocked? I would've thought that `jsdom` would already provide a suitable implementation of the function, and it adds an extra mock to clean up between tests.

File: src/line/i-chartBuilder.test.js [hunk 9]
```diff
@@ -0,0 +1,400 @@
...
+    const clearButton = buttons[1]
+
+    //make sure the correct button is selected
+    expect(clearButton).toHaveTextContent("Clear chart data")
+
+    const user = userEvent.setup()
+
+    await user.type(chartTitle, "Random Chart Title")
+    await user.type(xLabel, "X Label Name :)")
+    await user.type(yLabel, "Y Label Name :)")
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:274]: Based on the test name, I would expect this test to ensure the color input and X/Y input pairs are cleared when the "Clear chart data" button is pressed. Could you please add that?

File: src/line/i-chartBuilder.test.js [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+    expect(xValues).toHaveLength(1)
+    expect(yValues).toHaveLength(1)
+
+    const user = userEvent.setup()
+
+    await user.type(chartTitle, "Random Chart Title")
+    await user.type(xLabel, "X Label Name :)")
+    await user.type(yLabel, "Y Label Name :)")
+    await user.type(xValue, "22")
+    await user.type(yValue, "33")
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:327]: I think it would be great if data was entered for all of the X/Y input pairs, providing a stronger assertion that the X/Y input elements that exist after the "Clear chart data" button is pressed will be empty.

File: src/line/i-chartBuilder.test.js [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+    await user.click(clearButton)
+
+    xValues = domTesting.getAllByLabelText(document, "X")
+    yValues = domTesting.getAllByLabelText(document, "Y")
+
+    expect(xLabel).toHaveValue("")
+    expect(yLabel).toHaveValue("")
+    expect(chartTitle).toHaveValue("")
+    expect(xValues).toHaveLength(1)
+    expect(yValues).toHaveLength(1)
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:358]: Could you add assertions here that the X/Y input elements are empty?

File: src/line/i-chartBuilder.test.js [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+})
+
+//DATA CORRECTLY SENT TO CHART GENERATION FUNCTION
+test('generateChartImg receives the required parameters when the user creates a chart', async function(){
```
Comment (radiantBear) [inline src/line/i-chartBuilder.test.js:362]: I don't think I should be the one to review these last two tests since I wrote them. :)

**Review by radiantBear — CHANGES_REQUESTED** [review]
Thanks for writing the integration tests Noah! I left a few comments for small things I think would help guarantee the tests meet all requirements. 

I do have one bigger note as well - I think the way buttons are selected should be updated to decrease the brittleness of the tests. I'm of course open to the rest of the group's input, but I think using `getByText()` queries would make the tests more robust.

File: .rmv/config/db.mvbasic.json
Comment (radiantBear) [inline .rmv/config/db.mvbasic.json:1]: I don't think we need to have this file in the repo. Do your tests depend on this file? If not, could you remove it please?

File: .rmv/config/format.mvbasic.json
Comment (radiantBear) [inline .rmv/config/format.mvbasic.json:1]: Could you remove this file as well please?

File: .rmv/config/log4j2.properties
Comment (radiantBear) [inline .rmv/config/log4j2.properties:1]: Same with this file.

File: package.json [hunk 13]
```diff
@@ -1,37 +1,40 @@
...
+{
+  "name": "final-project",
+  "version": "1.0.0",
+  "description": "CS 362 Final Project",
+  "scripts": {
+    "build": "webpack",
+    "start": "webpack serve --open --port 3000",
+    "cy:open": "cypress open",
+    "cy:run": "cypress run",
+    "test": "jest",
```
Comment (radiantBear) [inline package.json]: If we define a general test script, it should probably also include the Cypress tests. Could you update this to call both or remove it please?

File: src/line/i-chartBuilder.test.js (hunk 12 above)
Comment (Rohanthap) [reply src/line/i-chartBuilder.test.js:362]: These look good to me!

File: .rmv/logs/rocket_mv_basic_language_server.log
Comment (radiantBear) [inline .rmv/logs/rocket_mv_basic_language_server.log:1]: I'm pretty sure this file doesn't need to be checked into version control.

File: .rmv/documentation/UNIVERSE/custom.json
Comment (radiantBear) [inline .rmv/documentation/UNIVERSE/custom.json:1]: Could you remove this file as well please?

File: src/line/i-chartBuilder.test.js (hunk 1 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js]: I still see different indentation on this line; could you please update it?

File: src/line/i-chartBuilder.test.js (hunk 4 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:49]: It still looks like we need an additional test here to assert that values entered into the X/Y coordinate inputs are preserved.

File: src/line/i-chartBuilder.test.js (hunk 7 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:117]: Could you please add this to the tests that define mocks/ spies?

File: src/line/i-chartBuilder.test.js (hunk 10 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:327]: Could you add this in please?

**Review by radiantBear — CHANGES_REQUESTED** [review]
Could you please reply to the open comments in the "Files changed" tab?

Also, I think `.rmv` should be added to the `.gitignore` and the files in that directory removed from version control.

File: src/line/i-chartBuilder.test.js (hunk 1 above)
Comment (noahdude1203) [reply src/line/i-chartBuilder.test.js]: I'm not seeing any indentation

File: .rmv/config/db.mvbasic.json
Comment (noahdude1203) [reply .rmv/config/db.mvbasic.json:1]: I don't know what this file is?

File: package.json (hunk 13 above)
Comment (noahdude1203) [reply package.json]: ?

Comment (radiantBear) [reply package.json]: You defined
```json
"test": "jest",
```
here. I'm saying if we are going to add the ability to run `npm test`, it should probably execute all the tests, including the Cypress ones. I think it would be better to just remove this line from the `scripts` object. 

Alternatively, one of the lecture repos had an additional dependency to allow running several commands from one npm `script`. I think it was called `npm-run-all` or something? We'd have to track it down, but then we could set `npm test` to execute both `jest` and `cypress run`.

File: src/line/i-chartBuilder.test.js (hunk 1 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js]: Sorry, line 38 - the one right above this comment.

File: src/line/i-chartBuilder.test.js (hunk 4 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:49]: Do you need help implementing this test?

File: src/line/i-chartBuilder.test.js (hunk 7 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:117]: Do you need help restoring the mocked functions?

File: src/line/i-chartBuilder.test.js (hunk 6 above)
Comment (radiantBear) [reply src/line/i-chartBuilder.test.js:118]: Do you need help implementing this test?

**Review by radiantBear — CHANGES_REQUESTED** [review]
It still looks like several required integration tests are missing. Do you need help writing them @noahdude1203?

**Review by radiantBear — APPROVED** [review]
Ready to merge
